### DISCUSSION TITLE
Ammending commits and other Git features in the web editor(`github.dev`)

### DISCUSSION THREAD

**Discussion by unrealapex** [community/community#8891] [Codespaces]
I feel that `github.dev` lacks some of the Git functionality that native VS Code has like amending commits for example, but I learned that there were technical and practical reasons for this feature not being added(discussed [here](https://github.com/github/feedback/discussions/5184#discussioncomment-1809020)). It would be nice if `github.dev` had close to or almost the same about of Git integration that native VS Code has. Amending commits would be useful if the web editor didn't push every time you commited(having an option to enable or disable automatic pushing, if disabled, user needs to push manually).

**Comment by bamurtaugh** [comment]
Thanks for opening, @UnrealApex! I'd love to know if there are specific motivations / common scenarios for when you amend commits in github.dev? Wondering if there might be a common thread that it is not apparent that commits get published to the remote automatically in github.dev, and folks are looking to amend their commits because of this.

**Reply by unrealapex** [reply]
>  Wondering if there might be a common thread that it is not apparent that commits get published to the remote automatically in github.dev

I think that might be the case, maybe adding a reminder somewhere in the UI that notifies users that pushing is automatic like "`Commits automatically pushed to repo`" would be helpful.
> Wondering if there might be a common thread that it is not apparent that commits get published to the remote automatically in github.dev

 Some common scenarios that require amending that most `github.dev` users have ran into are are misspelling a commit message, or forgetting to add a file to a commit before committing(main problem for me is fixing commit message typos).

**Reply by bamurtaugh** [reply]
Thank you for these insights! cc @joyceerhl too

**Reply by joyceerhl** [reply]
FWIW, we do already have a reminder in the source control box which indicates that commits go straight to the remote:

![image](https://user-images.githubusercontent.com/30305945/148265257-ec097ea9-a4f6-4dcf-b570-8a6b2c159a81.png)

Perhaps the message needs to be more explicit?

**Reply by unrealapex** [reply]
Forgot about that, thanks for pointing it out! I think that would be a good idea.

**Reply by ArwynFr** [reply]
Besides fixing typos and missing files, my main usage of `commit --amend` is when working with workflow.
This is usually a try and fail process which can produce a lot of commits if not amended.

**Reply by JohnyWS** [reply]
Another usecase is working in forks, where you want your commits to be cleaner. So while they are pushed, they are currently your own, until you open PR to the original repo (unless you use non-private repo yourself, and somebody else wants to work from your fork - not my problem though ;) ).

**Reply by karlhorky** [reply]
@joyceerhl @bamurtaugh any plans to add commit amending or any other Git features in 2024 / 2025?

If there's a different place that we can follow along other than this discussion, that would be great to know about too.

**Reply by bamurtaugh** [reply]
Thanks for following up on this! cc @lszomoru for any other thoughts, including if opening an issue to track something like this would be helpful.

**Comment by fulldecent** [comment]
Manual push doesn't make sense for me.

A git repository lives somewhere. Where would unpushed commits live?

If each commit needs to settle on the GitHub hosted repository, then this is a clean break—you are playing in your web browser, if you close it everything is gone, if you commit something that saves to the obvious place.

---

Amending is another story, amending is just a soft reset and a commit. This workflow also includes squashing. And this also fits cleanly with what github.dev already does. If it doesn't do that, I expect it will at some point without any controversy.

**Reply by joyceerhl** [reply]
> you are playing in your web browser, if you close it everything is gone

FYI, your uncommitted changes in github.dev are stored in local browser storage, so they are still there when you close and later revisit a repo in github.dev.

> amending is just a soft reset and a commit

In a traditional git workflow, the amended commit would still need to be force-pushed to the remote, so in a github.dev context we would be supporting `Commit Staged & Force Push (Staged)` rather than `Commit Staged (Amend)`.

**Comment by ItsHarper** [comment]
The primary situation where I find myself wanting this feature is when I make a change from the single-file editor, and then want to go back and edit that commit to also change other files as well.
